### DISCUSSION TITLE
Llama Ignoring Reverse Prompt Every Other Time

### DISCUSSION THREAD

**Discussion by loukylor** [ggml-org/llama.cpp#1200] [Q&A]
Hi, I'm sure this is a relatively easy fix, but I have an issue where the AI will ignore the reverse prompt in interact mode every other time:
![image](https://user-images.githubusercontent.com/46853114/234761502-efdfc006-b95f-4615-a42d-a68d265b0ef7.png)
You can see in the image it writes `User:` (which is my reverse prompt) but still continues to write without any input from me, where it responds again, then finally stops when the prompt comes up.

Here is my full command line: `main -m ./models/13B/ggml-model-q4_0.bin -n -1 --repeat_penalty 1.0 --color -i -r "User:" -f prompts/chat-with-bob.txt --in-prefix " " -c 2048 -t 8` 
The prompt is the `chat-with-bob.txt` from examples.

**Comment by klosax** [comment]
I also had this issue, but got it to work better if I didnt use the --in-prefix parameter and skipped the whitespace following the names. It seems that the reverse-prompting wont work well ending with whitespaces.
```
User:Hello, Bob.
Bob:Hello. How may I help you today?
User:Please tell me the largest city in Europe.
Bob:Sure. The largest city in Europe is Moscow, the capital of Russia.
User:
```

**Reply by loukylor** [reply]
According to the [docs](https://github.com/ggerganov/llama.cpp/tree/master/examples/main#in-prefix), I'm using the `--in-prefix` param properly, so perhaps this is an actual bug and not just incorrect usage?

**Comment by ivanstepanovftw** [comment]
Please, share full output including seed, so we can reproduce it

**Reply by loukylor** [reply]
Yep sorry bout that, here's another output I just generated.

Just because color isn't transferred, the only input I gave was `whats the tallest tower`. Everything before that was the prompt and everything after was generated.
```
E:\Code\AI\llama.cpp>main -m ./models/13B/ggml-model-q4_0.bin -n -1 --repeat_penalty 1.0 --color -i -r "User:" -f prompts/chat-with-bob.txt --in-prefix " " -c 2048 -t 8
main: seed = 1682652157
llama.cpp: loading model from ./models/13B/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v1 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 2048
llama_model_load_internal: n_embd     = 5120
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 40
llama_model_load_internal: n_layer    = 40
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 13824
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 13B
llama_model_load_internal: ggml ctx size =  73.73 KB
llama_model_load_internal: mem required  = 9807.47 MB (+ 1608.00 MB per state)
llama_init_from_file: kv self size  = 1600.00 MB

system_info: n_threads = 8 / 8 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | VSX = 0 |
main: interactive mode on.
Reverse prompt: 'User:'
Input prefix: ' '
sampling: temp = 0.800000, top_k = 40, top_p = 0.950000, repeat_last_n = 64, repeat_penalty = 1.000000
generate: n_ctx = 2048, n_batch = 512, n_predict = -1, n_keep = 0


== Running in interactive mode. ==
 - Press Ctrl+C to interject at any time.
 - Press Return to return control to LLaMa.
 - If you want to submit another line, end your input in '\'.

 Transcript of a dialog, where the User interacts with an Assistant named Bob. Bob is helpful, kind, honest, good at writing, and never fails to answer the User's requests immediately and with precision.

User: Hello, Bob.
Bob: Hello. How may I help you today?
User: Please tell me the largest city in Europe.
Bob: Sure. The largest city in Europe is Moscow, the capital of Russia.
User: whats the tallest tower
Bob: The tallest tower in Europe is the Ostankino Tower. It's 534.3 metres high. It's located in Moscow, the capital of Russia.
User: Bob: Hello.
User:
```

**Reply by DannyDaemonic** [reply]
The problem with your output is that model quantization has changed many times, it's hard to get a matching q4_0. It depends on which version you quantized it with. I tried a couple different q_4_0.bins I had for 13B and none of them matched.

Can you either requantize your `ggml-model-q4_0.bin` - as an added bonus, it will perform more accurately if you do - or just use the full f16.bin? Maybe try the 7B version if you're short on ram.

**Reply by ivanstepanovftw** [reply]
There is definetely something wrong with antiprompt, not with the model itself

**Reply by DannyDaemonic** [reply]
I'm just saying I can't reproduce that output using the given seed and model because my quantization of 13B is different than theirs.

**Reply by loukylor** [reply]
I just completely redownloaded the 13B and 7B models as well as requantize them and pull and build the latest llama.cpp and am still encountering the issue. So I think it is just something wrong with reverse prompt. Should I make an issue on the repo?

**Reply by CRD716** [reply]
@loukylor yeah, turn this into an issue, i've had this happen to me too.

**Reply by loukylor** [reply]
See #1224 

@klosax (just because I don't believe you'll be notified of this)

**Reply by DannyDaemonic** [reply]
> I just completely redownloaded the 13B and 7B models

Sorry. You didn't have to do that! To rule out corruption, there's a checksum included in the source, you could have compared it to the checksums there. 

> So I think it is just something wrong with reverse prompt. Should I make an issue on the repo?

Yeah, be sure to include a sample output made from a model that matches the checksums. If your quantized versions don't match use an `f16.bin`. (I don't know why none of my quantized models match any of the ones in the [sha256sums](https://github.com/ggerganov/llama.cpp/blob/master/SHA256SUMS) file. I rebuilt them recently.)

Might be a good idea to include what OS you're on, because I'm having trouble repeating this on Linux. I don't know how to show what version you're using but maybe just update to the latest version to ensure it's still happening.

**Reply by loukylor** [reply]
I include the info in the issue I just made, but I'm on Windows 10 and I'm on the latest commit (as of about 30 minutes ago).

**Reply by DannyDaemonic** [reply]
I just saw it. Thanks for posting. It looks good. Probably not too important, but they use sha256sum for the checksums, not md5sum.

**Reply by loukylor** [reply]
Oh, I just followed what was in the issue template. Might be worth it to pr to change it to sha256sum. Also thanks, no problem.

**Reply by DannyDaemonic** [reply]
Oh yeah, thanks for reminding me about that. I prefer `md5sum` myself but I'm guessing they've switched over to using `sha256sum` since that's the happyface standard and a lot of people are just downloading the models from there. I'll put a PR in to change it and see what they say.
